### ISSUE TITLE
Contradictions with context constructors

### ISSUE THREAD

**Issue by gmlueck** [KhronosGroup/SYCL-Docs#470]
I discovered this problem when reviewing work on the SYCL Reference guide (KhronosGroup/SYCL_Reference#73).  I'd like to get a quick response from the WG on this, so that the reference guide can document the correct information.

The SYCL 2020 specification of the `context` constructors is contradictory.  The [synopsis](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_context_interface) lists these constructors:

```
explicit context(const property_list& propList = {});

explicit context(async_handler asyncHandler,
                 const property_list& propList = {});

explicit context(const device& dev, const property_list& propList = {});

explicit context(const device& dev, async_handler asyncHandler,
                 const property_list& propList = {});

explicit context(const std::vector<device>& deviceList,
                 const property_list& propList = {});

explicit context(const std::vector<device>& deviceList,
                 async_handler asyncHandler,
                 const property_list& propList = {});
```

While [Table 19](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#table.constructors.context) lists these constructors:

```
explicit context(async_handler asyncHandler = {});

explicit context(const device& dev, async_handler asyncHandler = {});

explicit context(const std::vector<device>& deviceList,
                 async_handler asyncHandler = {});
```

Looking at the history of the specification, I see this same contradiction in SYCL 1.2.1.

I also noticed that we mysteriously removed two constructors that used to exist in SYCL 1.2.1:

```
context(const platform &plt, const property_list &propList = {});

context(const platform &plt, async_handler asyncHandler, const property_list &propList = {});
```

I can't think of any reason why we would want to remove these constructors, and there is no mention about removing them in the [What's changed](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:what-changed-between) chapter.  Therefore, I think this was probably just a mistake on our part.  Looking at the history, I see that these constructors were also removed (for no apparent reason) in the SYCL 2020 provisional specification.

I'd like to fix the SYCL 2020 specification to fix the contradiction and restore the two constructors that were removed from SYCL 1.2.1.  I believe the constructors listed in the synopsis are the correct ones, and the ones listed in Table 19 are wrong.  Therefore, I think this should be the complete list of `context` constructors in SYCL 2020:

```
explicit context(const property_list& propList = {});

explicit context(async_handler asyncHandler,
                 const property_list& propList = {});

explicit context(const device& dev, const property_list& propList = {});

explicit context(const device& dev, async_handler asyncHandler,
                 const property_list& propList = {});

explicit context(const platform &plt, const property_list &propList = {});

explicit context(const platform &plt, async_handler asyncHandler, const property_list &propList = {});

explicit context(const std::vector<device>& deviceList,
                 const property_list& propList = {});

explicit context(const std::vector<device>& deviceList,
                 async_handler asyncHandler,
                 const property_list& propList = {});
```

If there are no objections, I volunteer to fix the spec.

**Comment by TApplencourt** [comment]
Agree about fixing the inconsistency between the table and synopsis (ofc!)

Concerning the platform,
``
context(const platform &plt, const property_list &propList = {});
context(const platform &plt, async_handler asyncHandler, const property_list &propList = {});
```
Seems to be just syntactic sugar over
```
context(const platform plt.get_devices(), const property_list &propList = {});
context(const platform plt.get_devices(), async_handler asyncHandler, const property_list &propList = {});
```
Maybe the goal was to trim down the API surface?  Putting them back adds to the complexity of the `empty platform` proposal. Regardless, I have no strong opinion. Maybe A little preference of putting them back, just for backward compatibility

**Comment by gmlueck** [comment]
Yes, I also think we should restore those two constructors for backwards compatibility.  I think our intention with SYCL 2020 was to retain backward compatibility with SYCL 1.2.1 unless there was some overriding reason that required breaking compatibility.  I don't see any overriding reason in this case.
